src/Microsoft.SqlTools.ServiceLayer/LanguageServices/Completion/CompletionService.cs

```diff
@@ -0,0 +1,172 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.SqlServer.Management.SqlParser.Intellisense;
+using Microsoft.SqlServer.Management.SqlParser.MetadataProvider;
+using Microsoft.SqlServer.Management.SqlParser.Parser;
+using Microsoft.SqlTools.ServiceLayer.Connection;
+using Microsoft.SqlTools.ServiceLayer.LanguageServices.Contracts;
+
+namespace Microsoft.SqlTools.ServiceLayer.LanguageServices.Completion
+{
+    /// <summary>
+    /// A service to create auto complete list for given script document 
+    /// </summary>
+    internal class CompletionService
+    {
+        /// <summary>
+        /// Created new instance given binding queue
+        /// </summary>
+        public CompletionService(ConnectedBindingQueue bindingQueue)
+        {
+            BindingQueue = bindingQueue;
+        }
+
+        private ISqlParserWrapper sqlParserWrapper;
+
+        /// <summary>
+        /// SQL parser wrapper to create the completion list
+        /// </summary>
+        public ISqlParserWrapper SqlParserWrapper
+        {
+            get
+            {
+                if(this.sqlParserWrapper == null)
+                {
+                    this.sqlParserWrapper = new SqlParserWrapper();
+                }
+                return this.sqlParserWrapper;
+            }
+            set
+            {
+                this.sqlParserWrapper = value;
+            }
+        }
+
+        /// <summary>
+        /// Creates a completion list given connection and document info
+        /// </summary>
+        public AutoCompletionResult CreateCompletions(
+            ConnectionInfo connInfo,
+            ScriptDocumentInfo scriptDocumentInfo,
+            bool useLowerCaseSuggestions)
+        {
+            AutoCompletionResult result = new AutoCompletionResult();
+            // check if the file is connected and the file lock is available
+            if (scriptDocumentInfo.ScriptParseInfo.IsConnected && Monitor.TryEnter(scriptDocumentInfo.ScriptParseInfo.BuildingMetadataLock))
+            {
+                try
+                {
+                    QueueItem queueItem = AddToQueue(connInfo, scriptDocumentInfo.ScriptParseInfo, scriptDocumentInfo, useLowerCaseSuggestions);
+
+                    // wait for the queue item
+                    queueItem.ItemProcessed.WaitOne();
+                    var completionResult = queueItem.GetResultAsT<AutoCompletionResult>();
+                    if (completionResult != null && completionResult.CompletionItems != null)
+                    {
+                        result = completionResult;
+                    }
+                    else if (!ShouldShowCompletionList(scriptDocumentInfo.Token))
+                    {
+                        result.CompleteResult(scriptDocumentInfo.ScriptParseInfo.CurrentSuggestions, AutoCompleteHelper.EmptyCompletionList);
```

Comment:
actually I don't need it I'll remove it 

---
In reply to: [91597181](https://github.com/Microsoft/sqltoolsservice/pull/167#discussion_r91597181) [](ancestors = 91597181)